backend/src/main/java/ai/verta/modeldb/versioning/FindRepositoriesQuery.java

```diff
@@ -0,0 +1,355 @@
+            (key, value) -> {
+              if (value instanceof List) {
+                List<Object> objectList = (List<Object>) value;
+                query.setParameterList(key, objectList);
+              } else {
+                query.setParameter(key, value);
+              }
+            });
+      }
+    }
+
+    public Query buildCountQuery() {
+      Query query = session.createQuery(this.countQueryString);
+      setParameterInQuery(query);
+      return query;
+    }
+
+    private String getHQLQueryString() {
+      boolean isNeedOperator = false;
+      String alias = " repo";
+      StringBuilder queryBuilder =
+          new StringBuilder(" FROM ").append(RepositoryEntity.class.getSimpleName()).append(alias);
+
+      StringBuilder joinClause = new StringBuilder();
+      if (this.predicates != null && !this.predicates.isEmpty()) {
+        this.predicates.forEach(
+            keyValueQuery -> {
+              if (keyValueQuery.getKey().contains(ModelDBConstants.LABEL)) {
+                joinClause
+                    .append(" INNER JOIN ")
+                    .append(LabelsMappingEntity.class.getSimpleName())
+                    .append(" ")
+                    .append(ModelDBConstants.LABELS)
+                    .append(" ON ");
+                joinClause
+                    .append(ModelDBConstants.LABELS)
+                    .append(".id.entity_hash = ")
+                    .append(alias)
+                    .append(".")
+                    .append(ModelDBConstants.ID)
+                    .append(" AND ");
+                joinClause
+                    .append(ModelDBConstants.LABELS)
+                    .append(".id.entity_type = ")
+                    .append(IDTypeEnum.IDType.VERSIONING_REPOSITORY.getNumber());
+              }
+            });
+      }
+
+      StringBuilder whereClause = new StringBuilder();
+      if (workspaceDTO != null
+          && workspaceDTO.getWorkspaceId() != null
+          && !workspaceDTO.getWorkspaceId().isEmpty()) {
+        whereClause.append(" WHERE ");
+        whereClause
+            .append(alias)
+            .append(".")
+            .append(ModelDBConstants.WORKSPACE_ID)
+            .append(" = :")
+            .append(ModelDBConstants.WORKSPACE_ID);
+        parametersMap.put(ModelDBConstants.WORKSPACE_ID, workspaceDTO.getWorkspaceId());
+        whereClause
+            .append(" AND ")
+            .append(alias)
+            .append(".")
+            .append(ModelDBConstants.WORKSPACE_TYPE)
+            .append(" = :")
+            .append(ModelDBConstants.WORKSPACE_TYPE);
+        parametersMap.put(
+            ModelDBConstants.WORKSPACE_TYPE, workspaceDTO.getWorkspaceType().getNumber());
+        isNeedOperator = true;
+      } else if (this.isNeedToWhereCause) {
+        whereClause.append(" WHERE ");
+      }
+
+      if (this.predicates != null && !this.predicates.isEmpty()) {
+        if (isNeedOperator) {
+          whereClause.append(" AND ");
```

Comment:
Rather than this AND sprinkled at multiple places can we collect predicate strings in an array and then have a function to append them with AND?
it should take the join string as parameter so that we can use the same to construct an or condition.

Reply:
Added